index.js

```diff
@@ -0,0 +1,98 @@
+//.
+//. Tiny and framework agnostic role-based permission management using
+//. a model of *composition* over an inheritance-based model.
+//.
+//. > `npm install --save permissionary`
+//.
+//. ## Usage
+//.
+//. ```js
+//. var permissionary = require('permissionary');
+//. ```
+//.
+//. * [Basic](#permissionary)
+//. * [With Express][1]
+//.
+//. ## Philosophy
+//.
+//. Many permission systems use the idea of inheritance to define roles in
+//. terms of other roles. This can lead to the definition of non-flexible
+//. roles where the developer has to make decisions that determine what will
+//. be possible in the future. [Mattias Petter Johansson][3] has
+//. [a good video][4] explaining the phenomenon.
+//.
+//. To combat this issue, Permissionary has no inheritance. Instead, groupings
+//. of grants are given names (known as roles), and multiple such roles can be
+//. assigned to a user. This allows one to define a
+//.
+//. ## API
+
+'use strict';
+
+var micromatch = require('micromatch');
+
+//# permissionary :: StrMap (Array String) -> Array String -> String -> Boolean
+//.
+//. A [curried][5] function that takes three arguments and returns a Boolean:
+//.
+//. 1. A mapping from roles to an array of glob patterns.
+//. 2. An Array of roles.
+//. 3. A permission.
+//.
+//. The glob patterns will be filtered down to contain only those associated
+//. with the given list of roles. The permission will be checked against the
+//. filtered glob patters using [micromatch][6] to produce the Boolean.
+//.
+//. To make optimal use of this function, it is recommended to partially apply
+//. the function to produce new functions, as shown in the example below:
+//.
+//. ```js
+//. // This defines a mapping from roles to permissions.
+//. // We can use wildcards to assign multiple permissions at once.
+//. var createVerifier = permissionary({
+//.   'content-reader': ['content.read', 'images.read'],
+//.   'content-writer': ['content.write', 'images.upload'],
+//.   'superadmin':     ['*']
+//. });
+//.
+//. // Let's say our user Bob is a content-reader, and also a content-writer.
+//. var canBob = createVerifier(['content-reader', 'content-writer']);
+//.
+//. // And Alice is an administrator.
+//. var canAlice = createVerifier(['superadmin']);
+//.
+//. // This will return true, because Bob has this permission
+//. // through his content-reader role.
+//. canBob('content.read');
+//.
+//. // This will return false, bob does not have this permission.
+//. canBob('users.create');
+//.
+//. // Alice, however, does. She has all permissions (even the ones
+//. // we haven't thought of yet), so this will return true.
+//. canAlice('users.create');
+//. ```
+module.exports = function permissionary(grants, groups, permission) {
+
+  if (arguments.length === 1) {
+    return permissionary.bind(this, grants);
+  }
+
+  var permissions = groups.reduce(function(permissions, group) {
+    return grants[group] ? permissions.concat(grants[group]) : permissions;
```

Comment:
This will produce false positives if `group` is `'toString'` or `'valueOf'`. Consider `Object#hasOwnProperty`.